### PR TITLE
Bug fixes from migration run using Slingshot files

### PR DESCRIPTION
**Pull request by guymaness2**
**Thank you for your code contribution, we will take your code suggestion into consideration. It will be reviewed, and based on code quality and need addressed, we will determine acceptance. If this code addresses a feature request or issue, be sure to link to that issue.** 

##### What does the change add or fix?

* Various bug fixes to allow multiple bulldozer passes on same data base without dictionary conflicts.
* Various changes and fixes to be able to support using PCO Slingshot generated csv files for import.

---------

##### What does the change add or fix in a succinct statement that will be read by clients?

* Various bug fixes to allow multiple bulldozer passes on same data base without dictionary conflicts.
* Various changes and fixes to be able to support using PCO Slingshot generated csv files for import.

---------

##### Who reported, requested, or paid for the change?

Warranty

---------

##### Does this update or add options to the block UI?

no

---------

##### What files does it affect?

* Bulldozer.CSV/Bulldozer.CSV.csproj
* Bulldozer.CSV/CSVComponent.cs
* Bulldozer.CSV/CSVInstance.cs
* Bulldozer.CSV/Maps/Attendance.cs
* Bulldozer.CSV/Maps/Financial.cs
* Bulldozer.CSV/Maps/FinancialTransaction.cs
* Bulldozer.CSV/Maps/Group.cs
* Bulldozer.CSV/Maps/Location.cs
* Bulldozer.CSV/Maps/Notes.cs
* Bulldozer.CSV/Maps/Person.cs
* Bulldozer.CSV/Maps/Prayer.cs
* Bulldozer.CSV/Maps/Schedule.cs
* Bulldozer.CSV/Model/EntityNoteCsv.cs
* Bulldozer.CSV/Model/FinancialAccountCsv.cs
* Bulldozer.CSV/Model/FinancialBatchCsv.cs
* Bulldozer.CSV/Model/LocationCsv.cs
* Bulldozer.CSV/Model/PersonNoteCsv.cs
* Bulldozer.CSV/Model/ScheduleCsv.cs
* Bulldozer/Bulldozer.csproj
* Bulldozer/Model/AttendanceImport.cs
* Bulldozer/Model/NoteImport.cs
* Bulldozer/Utility/AddMethods.cs
* Bulldozer/Utility/Extensions.cs  
  
---------

##### Is it a breaking change for other versions/clients?

no

### CODE REVIEW

File: Bulldozer.CSV/Maps/Location.cs [hunk 1]
```diff
@@ -182,5 +182,95 @@ private void SaveNamedLocation( List<Location> newNamedLocationList )
...
+                    Street1 = locationCsv.Street1.Left( 100 ),
+                    Street2 = locationCsv.Street2.Left( 100 ),
+                    City = locationCsv.City.Left( 50 ),
+                    County = locationCsv.County.Left( 50 ),
+                    State = locationCsv.State.Left( 50 ),
+                    Country = locationCsv.Country.Left( 50 ),
+                    PostalCode = locationCsv.PostalCode.Left( 50 ),
+                    CreatedDateTime = importedDateTime,
+                    ModifiedDateTime = importedDateTime,
+                    ForeignKey = $"{this.ImportInstanceFKPrefix}^{locationCsv.Id}",
```
Comment (guymaness2) [inline Bulldozer.CSV/Maps/Location.cs:222]:
```suggestion
                    ForeignKey = foreignKey,
```

File: Bulldozer.CSV/Maps/Schedule.cs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+                var newSchedule = new Schedule
+                {
+                    Name = scheduleCsv.Name.Left( 50 ),
+                    WeeklyDayOfWeek = scheduleCsv.MeetingDay,
+                    WeeklyTimeOfDay = scheduleCsv.MeetingTime,
+                    Description = scheduleCsv.Description,
+                    CreatedDateTime = importedDateTime,
+                    ModifiedDateTime = importedDateTime,
+                    ForeignKey = $"{this.ImportInstanceFKPrefix}^{scheduleCsv.Id}",
```
Comment (nateh777) [inline Bulldozer.CSV/Maps/Schedule.cs:68]:
```suggestion
                    ForeignKey = foreignKey,
```

File: Bulldozer.CSV/CSVComponent.cs [hunk 3]
```diff
@@ -624,9 +656,9 @@ public override int TransformData( Dictionary<string, string> settings )
             //// Insert Financial related data
 
             var financialbatchInstance = selectedCsvData.FirstOrDefault( i => i.RecordType == CSVInstance.RockDataType.BATCH );
```
Comment (nateh777) [inline Bulldozer.CSV/CSVComponent.cs:658]:
```suggestion
```

File: Bulldozer.CSV/CSVComponent.cs [hunk 4]
```diff
@@ -1354,29 +1390,69 @@ private void LoadImportEntityLists( List<CSVInstance> csvInstances )
...
 
             // Groups (non-family)
             LoadGroupDataLists( csvInstances );
 
+            // Financial Accounts
+            var financialAccountInstance = csvInstances.FirstOrDefault( i => i.RecordType == CSVInstance.RockDataType.FinancialAccount );
+            if ( financialAccountInstance != null )
+            {
+                FinancialAccountCsvList = LoadEntityImportListFromCsv<FinancialAccountCsv>( financialAccountInstance.FileName );
+                ReportProgress( 0, string.Format( "FinancialAcount records: {0}", FinancialAccountCsvList.Count ) );
```
Comment (nateh777) [inline Bulldozer.CSV/CSVComponent.cs:1421]:
```suggestion
                ReportProgress( 0, string.Format( "FinancialAccount records: {0}", FinancialAccountCsvList.Count ) );
```

File: Bulldozer.CSV/CSVInstance.cs [hunk 5]
```diff
@@ -139,6 +158,7 @@ public enum RockDataType
             RELATIONSHIP,
             ATTENDANCE,
             METRICS,
+            Locations,
```
Comment (nateh777) [inline Bulldozer.CSV/CSVInstance.cs:161]: This should be renamed to `Location` non plural.

File: Bulldozer.CSV/Maps/Financial.cs [hunk 6]
```diff
@@ -1123,5 +1124,212 @@ public FinancialAccount AddAccount( RockContext lookupContext, string fundName,
...
+                }
+                if ( account.PublicName.IsNotNullOrWhiteSpace() )
+                {
+                    newAccount.PublicName = account.PublicName.Truncate( 50 );
+                }
+                else
+                {
+                    newAccount.PublicName = accountName;
+                }
+
```
Comment (nateh777) [inline Bulldozer.CSV/Maps/Financial.cs:1182]:
```suggestion
```

File: Bulldozer.CSV/Maps/Financial.cs [hunk 7]
```diff
@@ -1123,5 +1124,212 @@ public FinancialAccount AddAccount( RockContext lookupContext, string fundName,
...
+                if ( completed > 0 && completed % ( this.DefaultChunkSize * 10 ) < 1 )
+                {
+                    ReportProgress( 0, $"{completed} Financial Batches processed." );
+                }
+
+                if ( completed % this.DefaultChunkSize < 1 )
+                {
+                    var csvChunk = workingBatchImportList.Take( Math.Min( this.DefaultChunkSize, workingBatchImportList.Count ) ).ToList();
+                    rockContext.BulkInsert( batchesToInsert );
+                    completed += insertedBatches.Count;
```
Comment (nateh777) [inline Bulldozer.CSV/Maps/Financial.cs:1325]: This is not chunking appropriately.

File: Bulldozer.CSV/Model/FinancialBatchCsv.cs [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+        public DateTime? EndDate { get; set; }
+
+        public BatchStatus Status { get; set; }
+
+        public string CreatedByPersonId { get; set; }
+
+        public DateTime? CreatedDateTime { get; set; }
+
+        public string ModifieddByPersonId { get; set; }
```
Comment (guymaness2) [inline Bulldozer.CSV/Model/FinancialBatchCsv.cs:25]: Typo

File: Bulldozer/Utility/AddMethods.cs [hunk 9]
```diff
@@ -1848,6 +1848,79 @@ public static int AddBenevolenceType( RockContext rockContext, string typeName )
...
+        /// Gets the campus identifier.
+        /// </summary>
+        /// <param name="property">Name of the property.</param>
+        /// <param name="includeCampusName">if set to <c>true</c> [include campus name].</param>
+        /// <returns></returns>
+        public static int? GetCampus( string campusIdString, string importPrefix, bool existingCampusesOnly = false, string possibleCampusName = null, bool addNew = false )
+        {
+            var rockContext = new RockContext();
+            int? returnCampusId = null;
+            Campus campus = null;
```
Comment (nateh777) [inline Bulldozer/Utility/AddMethods.cs:1861]: Is this necessary to be declared outside of the if statements? Looks like it is only used in one

File: Bulldozer.CSV/Maps/Notes.cs [hunk 10]
```diff
@@ -155,6 +156,205 @@ private static void SaveNotes( List<Note> noteList )
...
+        /// Maps the person note data.
+        /// </summary>
+        /// <exception cref="System.NotImplementedException"></exception>
+        private int ImportEntityNote<T>( IEnumerable<EntityNoteCsv> entityNoteList, bool? groupEntityIsFamily )
+        {
+            ReportProgress( 0, "Preparing PersonNote data for import..." );
+
+            var entityType = EntityTypeCache.Get<T>();
+            var rockContext = new RockContext();
+            var notesToInsert = new List<FinancialBatch>();
```
Comment (nateh777) [inline Bulldozer.CSV/Maps/Notes.cs:170]: Not used.

File: Bulldozer.CSV/Maps/Notes.cs [hunk 11]
```diff
@@ -155,6 +156,205 @@ private static void SaveNotes( List<Note> noteList )
...
+                .Where( a => a.EntityTypeId == entityType.Id ).Select( a => new
+                {
+                    a.Id,
+                    a.Name
+                } )
+                .ToList()
+                .DistinctBy( a => a.Name )
+                .ToDictionary( k => k.Name, v => v.Id, StringComparer.OrdinalIgnoreCase );
+
+            var importNoteTypeNames = this.PersonNoteCsvList.Select( a => a.NoteType ).Distinct().ToList();
```
Comment (guymaness2) [inline Bulldozer.CSV/Maps/Notes.cs:186]: Should be looking at entityNoteList being passed in.

File: Bulldozer.CSV/Maps/Notes.cs [hunk 12]
```diff
@@ -155,6 +156,205 @@ private static void SaveNotes( List<Note> noteList )
...
+        /// <exception cref="System.NotImplementedException"></exception>
+        private int ImportEntityNote<T>( IEnumerable<EntityNoteCsv> entityNoteList, bool? groupEntityIsFamily )
+        {
+            ReportProgress( 0, "Preparing PersonNote data for import..." );
+
+            var entityType = EntityTypeCache.Get<T>();
+            var rockContext = new RockContext();
+            var notesToInsert = new List<FinancialBatch>();
+            var noteImportList = new List<NoteImport>();
+            var personEntityType = EntityTypeCache.Get( PersonEntityTypeId );
```
Comment (guymaness2) [inline Bulldozer.CSV/Maps/Notes.cs:172]: should this be more generic "entityType"?

File: Bulldozer.CSV/Maps/Notes.cs [hunk 13]
```diff
@@ -155,6 +156,205 @@ private static void SaveNotes( List<Note> noteList )
...
+
+        /// <summary>
+        /// Bulks the note import.
+        /// </summary>
+        /// <param name="noteImports">The note imports.</param>
+        /// <param name="entityTypeId">The entity type identifier.</param>
+        /// <param name="foreignSystemKey">The foreign system key.</param>
+        /// <param name="groupEntityIsFamily">If this is a GroupEntity, is it a Family GroupType?</param>
+        /// <returns></returns>
+        public int BulkNoteImport( List<NoteImport> noteImports, int entityTypeId, bool? groupEntityIsFamily )
```
Comment (guymaness2) [inline Bulldozer.CSV/Maps/Notes.cs:275]: This code seems unfinished. Needs worked through by Guy before we review again

File: Bulldozer.CSV/Maps/Notes.cs [hunk 14]
```diff
@@ -155,6 +156,205 @@ private static void SaveNotes( List<Note> noteList )
                 rockContext.SaveChanges( DisableAuditing );
             } );
         }
+
+        /// <summary>
+        /// Maps the person note data.
+        /// </summary>
+        /// <exception cref="System.NotImplementedException"></exception>
+        private int ImportEntityNote<T>( IEnumerable<EntityNoteCsv> entityNoteList, bool? groupEntityIsFamily )
```
Comment (guymaness2) [inline Bulldozer.CSV/Maps/Notes.cs:164]: This code seems unfinished. Needs worked through by Guy before we review again

**Comment by guymaness2** [comment]
Ended on https://github.com/KingdomFirst/Bulldozer/pull/100/commits/67bacca4510b3479d82788c8d135d796f3db596b

**Comment by guymaness2** [comment]
@nateh777 resolved all issues from code review. Reworked and tested Notes import section. Ready for review.

File: Bulldozer.CSV/CSVComponent.cs [hunk 15]
```diff
@@ -2475,25 +2563,24 @@ private int AddPhoneCountryCodeDVs( RockContext rockContext = null )
...
             
             var attributeDefinedValuesDict = attributeDefinedTypeDict.ToDictionary( k => k.Key, v => v.Value.DefinedValues.ToDictionary( a => a.Value, b => b.DefinedTypeId ) );
 
-            var attributeValuesToProcess = attributeValues.Where( v => attributeDefinedValuesDict.ContainsKey( $"{v.EntityTypeId}_{v.AttributeKey}" ) && !attributeDefinedValuesDict[$"{v.EntityTypeId}_{v.AttributeKey}"].ContainsKey( v.AttributeValue  ) )
+            var attributeValuesToProcess = attributeValues.Where( v => attributeDefinedValuesDict.ContainsKey( $"{v.EntityTypeId}_{v.AttributeKey}_{v.EntityTypeQualifierValue}" ) && !attributeDefinedValuesDict[$"{v.EntityTypeId}_{v.AttributeKey}_{v.EntityTypeQualifierValue}"].ContainsKey( v.AttributeValue  ) )
                                                           .Select( v => new AttributeValueObject { AttributeKey = v.AttributeKey, AttributeValue = v.AttributeValue, EntityTypeId = v.EntityTypeId, DefinedTypeId = attributeDefinedTypeDict.GetValueOrNull( $"{v.EntityTypeId}_{v.AttributeKey}" )?.Id } )
```
Comment (nateh777) [inline Bulldozer.CSV/CSVComponent.cs:2584]: Does this line also need to include the qualifiervalue? I would think so. Be sure to check `attributeDefinedValuesDict` and `attributeValuesToProcess` need to be consistently using the same key value with the qualifiervalue.

```suggestion
                                                          .Select( v => new AttributeValueObject { AttributeKey = v.AttributeKey, AttributeValue = v.AttributeValue, EntityTypeId = v.EntityTypeId, DefinedTypeId = attributeDefinedTypeDict.GetValueOrNull( $"{v.EntityTypeId}_{v.AttributeKey}_{v.EntityTypeQualifierValue}" )?.Id } )
```

File: Bulldozer.CSV/Maps/Notes.cs [hunk 16]
```diff
@@ -155,6 +157,206 @@ private static void SaveNotes( List<Note> noteList )
...
+                        };
+
+                        noteImportList.Add( newNoteImport );
+                    }
+
+                    // Slice data into chunks and process
+                    var workingNoteImportList = noteImportList.ToList();
+                    var notesRemainingToProcess = noteImportList.ToList().Count;
+                    var completed = 0;
+                    var insertedNotes = new List<Note>();
```
Comment (nateh777) [inline Bulldozer.CSV/Maps/Notes.cs:258]:
```suggestion
```

**Comment by guymaness2** [comment]
@nateh777 pulled changes in and rebuilt. Was there something you needed me to test, or are we good for merge?

**Comment by nateh777** [comment]
@guymaness2 I believe just attributeDefinedTypeDict changes being checked to ensure we really are using the proper keys everywhere, otherwise I think it was good to go. I'll approve.

**Review by nateh777 — APPROVED** [review]

**Pull request merged**